parasol/query.py

```diff
@@ -0,0 +1,241 @@
+    searching and filtering of Solr results. Allows search results
+    to be paginated by django paginator."""
+
+    _result_cache = None
+    start = 0
+    stop = None
+    sort_options = []
+    search_qs = []
+    filter_qs = []
+
+    #: by default, combine search queries with AND
+    default_search_operator = 'AND'
+
+    def __init__(self, solr: Optional[SolrClient] = None):
+        self.solr = solr or SolrClient()
+        # convert search operator into form needed for combining queries
+        self._search_op = ' %s ' % self.default_search_operator
+
+    def get_results(self, **kwargs):
+        """
+        Query Solr and get the results for the current query and filter
+        options. Populates result cache and returns the documents portion
+        of the reponse.
+
+        Returns:
+            Solr response documents as a list of dictionaries.
+        """
+        query_opts = self.query_opts()
+        query_opts.update(**kwargs)
+        # TODO: what do we do about the fact that Solr defaults
+        # to 10 rows?
+
+        # NOTE: django templates choke on AttrDict because it is
+        # callable; using dictionary response instead
+        self._result_cache = self.solr.query(wrap=False, **query_opts)
+        return self._result_cache['docs']
+
+    def query_opts(self):
+        """Construct query options based on current queryset configuration.
+        Includes filter queries, start and rows, sort, and search query.
+        """
+        query_opts = {
+            'start': self.start,
+        }
+        if self.filter_qs:
+            query_opts['fq'] = self.filter_qs
+        if self.stop:
+            query_opts['rows'] = self.stop - self.start
+        if self.sort_options:
+            query_opts['sort'] = ','.join(self.sort_options)
+
+        # main query; if no query is defined, find everything
+        if self.search_qs:
+            query_opts['q'] = self._search_op.join(self.search_qs)
+        else:
+            query_opts['q'] = '*:*'
+
+        return query_opts
+
+    def count(self):
+        """Total number of results for the current query"""
+
+        # if result cache is already populated, use it
+        if self._result_cache is not None:
+            return self._result_cache['numFound']
+
+        # otherwise, query with current options but request zero rows
+        # and do not populate the result cache
+        query_opts = self.query_opts()
+        query_opts['rows'] = 0
+        return self.solr.query(**query_opts, wrap=False)['numFound']
+
+    @staticmethod
+    def _lookup_to_filter(key, value):
+        """Convert keyword argument key=value pair into a Solr filter.
+        Currently only supports simple case of field:value."""
+
+        # NOTE: as needed, we can start implementing django-style filters
+        # such as __in=[a, b, c] or __range=(start, end)
+        return '%s:%s' % (key, value)
+
+    def filter(self, *args, **kwargs):
```

Comment:
I think this one could benefit from some docstring on what to expect in terms of `args` `kwargs` format. Possibly just a restatement of the form given in the module docstring.